### PR TITLE
Unscheduled session

### LINKED ISSUE #290: Unscheduled sessions should be possible
Sessions represent the event owner's copy of a Proposal, allowing for customisation without changing the original. However, when scheduling events, @zainabbawa frequently finds herself moving proposals into and out of the schedule to find a correct fit. Since sessions require a mandatory start/end time, this means session objects are created and deleted on such moves, losing all customisation.

Instead, sessions should support an unscheduled mode and should not be deleted. This is necessary to preserve customisations and to support featuring (#289) before the schedule is finalised.

### PR DESCRIPTION
**Pull request by iambibhas**
Fixes #290. 

Retains session objects in database when sessions are deleted from schedule, and reuse them whenever a session for the same proposal is created again.

### CODE REVIEW

File: funnel/models/project.py [hunk 1]
```diff
@@ -88,7 +88,9 @@ class Project(BaseScopedNameMixin, db.Model):
     labels = db.Column(JsonDict, nullable=False, server_default='{}')
 
     featured_sessions = db.relationship("Session",
-        primaryjoin="and_(Session.project_id == Project.id, Session.featured == True)", back_populates="project")
+        primaryjoin="and_(Session.project_id == Project.id, Session.start != None, Session.end != None, Session.featured == True)", back_populates="project")
```
Comment (jace) [inline funnel/models/project.py]: Unscheduled sessions can be featured. That was the main motivation for having unscheduled sessions.

File: funnel/models/project.py [hunk 2]
```diff
@@ -88,7 +88,9 @@ class Project(BaseScopedNameMixin, db.Model):
...
+    scheduled_sessions = db.relationship("Session",
+        primaryjoin="and_(Session.project_id == Project.id, Session.start != None, Session.end != None)", back_populates="project")
```
Comment (jace) [inline funnel/models/project.py]: You'll also need one for unscheduled sessions.

File: funnel/models/session.py [hunk 3]
```diff
@@ -23,8 +23,8 @@ class Session(UuidMixin, BaseScopedIdNameMixin, db.Model):
     proposal = db.relationship(Proposal,
         backref=db.backref('session', uselist=False, cascade='all, delete-orphan'))
     speaker = db.Column(db.Unicode(200), default=None, nullable=True)
-    start = db.Column(db.DateTime, nullable=False)
-    end = db.Column(db.DateTime, nullable=False)
+    start = db.Column(db.DateTime, nullable=True)
+    end = db.Column(db.DateTime, nullable=True)
```
Comment (jace) [inline funnel/models/session.py]: Add a check constraint to ensure both or neither are set.

File: funnel/models/venue.py [hunk 4]
```diff
@@ -44,6 +44,9 @@ class VenueRoom(BaseScopedNameMixin, db.Model):
     description = MarkdownColumn('description', default=u'', nullable=False)
     bgcolor = db.Column(db.Unicode(6), nullable=False, default=u'229922')
 
+    scheduled_sessions = db.relationship("Session",
+        primaryjoin="and_(Session.venue_room_id == VenueRoom.id, Session.start != None, Session.end != None)", back_populates="venue_room")
```
Comment (jace) [inline funnel/models/venue.py]: Did you really mean to use `back_populates`? https://docs.sqlalchemy.org/en/latest/orm/relationship_api.html#sqlalchemy.orm.relationship.params.back_populates

File: funnel/templates/session_view_popup.html.jinja2 [hunk 5]
```diff
@@ -8,12 +8,18 @@
...
+    {% if session.scheduled %}
+      <p class="mui--text-subhead zero-bottom-margin">
+        <small class="mui--text-light">at</small>
+        {{ localize_date(session.start, to_tz=timezone).strftime('%I:%M %p') }} - {{ localize_date(session.end, to_tz=timezone).strftime('%I:%M %p') }}
+        <small class="mui--text-light">on</small>
+        {{ localize_date(session.start, to_tz=timezone).strftime('%d %B, %Y') }}
+      </p>
+    {% else %}
+      <p class="mui--text-subhead zero-bottom-margin">
+        This session has not been scheduled yet.
```
Comment (jace) [inline funnel/templates/session_view_popup.html.jinja2]: Add `{% trans %}`

File: funnel/views/schedule.py [hunk 6]
```diff
@@ -81,6 +81,12 @@ def schedule_data(project):
 
 
 def session_ical(session):
+    # This function is only called with scheduled sessions.
+    # If for some reason it is used somewhere else and called with an unscheduled session,
+    # this function should fail.
+    if not session.scheduled:
+        raise Exception(u"{} is not scheduled".format(session))
```
Comment (jace) [inline funnel/views/schedule.py]: Use `{0!r}` instead of `{}`.

Comment (jace) [reply funnel/views/schedule.py]: https://docs.python.org/2.7/library/string.html#format-string-syntax

File: funnel/views/schedule.py [hunk 7]
```diff
@@ -259,8 +265,10 @@ def schedule_edit(profile, project):
         'scheduled': session_data(project.sessions, timezone=project.timezone, with_modal_url='edit', with_delete_url=True)
         }
     # Set the proper range for the calendar to allow for date changes
-    first_session = Session.query.filter_by(project=project).order_by("created_at asc").first()
-    last_session = Session.query.filter_by(project=project).order_by("created_at desc").first()
+    first_session = Session.query.filter(
+        Session.start.isnot(None), Session.end.isnot(None)).filter_by(project=project).order_by("created_at asc").first()
+    last_session = Session.query.filter(
+        Session.start.isnot(None), Session.end.isnot(None)).filter_by(project=project).order_by("created_at desc").first()
```
Comment (jace) [inline funnel/views/schedule.py]: Make `scheduled` a `hybrid_property` and you can rewrite this as `Session.query.filter(Session.scheduled)…`

File: funnel/views/session.py [hunk 8]
```diff
@@ -17,15 +17,19 @@ def rooms_list(project):
 
 
 def session_form(project, proposal=None, session=None):
-    if session:
+    # Look for any existing unscheduled session
+    if proposal and not session:
+        session = Session.query.filter_by(project=project, proposal=proposal).first()
```
Comment (jace) [inline funnel/views/session.py]: Add a classmethod to Session and make this call `Session.for_proposal(proposal, create=True)`

File: funnel/views/session.py [hunk 9]
```diff
@@ -111,6 +115,9 @@ def session_edit(profile, project, session):
     permission='edit-session')
 def session_delete(profile, project, session):
     modal_url = session.proposal.url_for('schedule') if session.proposal else None
-    db.session.delete(session)
+    # Session is not deleted, but we remove start and end time,
+    # so it becomes an unscheduled session.
+    session.start = None
+    session.end = None
```
Comment (jace) [inline funnel/views/session.py]: Make this a method: `Session.remove_schedule()`

Comment (jace) [reply funnel/views/session.py]: Or `Session.make_unscheduled()`

File: funnel/models/venue.py (hunk 4 above)
Comment (iambibhas) [reply funnel/models/venue.py]: Let me test couple of things. The complementing property already exists for us.

File: funnel/models/session.py [hunk 10]
```diff
@@ -23,15 +24,42 @@ class Session(UuidMixin, BaseScopedIdNameMixin, db.Model):
...
     is_break = db.Column(db.Boolean, default=False, nullable=False)
     featured = db.Column(db.Boolean, default=False, nullable=False)
     banner_image_url = db.Column(db.Unicode(2000), nullable=True)
 
-    __table_args__ = (db.UniqueConstraint('project_id', 'url_id'),)
+    __table_args__ = (
+        db.UniqueConstraint('project_id', 'url_id'),
+        db.CheckConstraint(
+            u'("start" IS NULL AND "end" IS NULL) OR ("start" IS NOT NULL AND "end" IS NOT NULL)',
+            u'ck_session_start_end_nullable')
```
Comment (jace) [inline funnel/models/session.py]: Coding convention is `_check` suffix, so `session_start_end_check`.

File: funnel/models/session.py [hunk 11]
```diff
@@ -23,15 +24,42 @@ class Session(UuidMixin, BaseScopedIdNameMixin, db.Model):
...
+        )
+
+    @hybrid_property
+    def scheduled(self):
+        # A session is scheduled only when both start and end fields have value
+        return (self.start != None) & (self.end != None)
```
Comment (jace) [inline funnel/models/session.py]: You'll need separate implementations for Python and SQLAlchemy. This should be the SQL version defined with `@scheduled.expression`

File: funnel/models/session.py [hunk 12]
```diff
@@ -23,15 +24,42 @@ class Session(UuidMixin, BaseScopedIdNameMixin, db.Model):
...
+
+    @classmethod
+    def for_proposal(cls, proposal, create=False):
+        session_obj = cls.query.filter_by(proposal=proposal).first()
+        if session_obj is None and create:
+            session_obj = cls(title=proposal.title, description=proposal.description,
+                speaker_bio=proposal.bio, project=proposal.project, proposal=proposal   )
+            session_obj.make_id()
+            session_obj.make_name()
```
Comment (jace) [inline funnel/models/session.py]: This is now automatic.

File: funnel/views/schedule.py [hunk 13]
```diff
@@ -259,8 +265,8 @@ def schedule_edit(profile, project):
         'scheduled': session_data(project.sessions, timezone=project.timezone, with_modal_url='edit', with_delete_url=True)
         }
     # Set the proper range for the calendar to allow for date changes
-    first_session = Session.query.filter_by(project=project).order_by("created_at asc").first()
-    last_session = Session.query.filter_by(project=project).order_by("created_at desc").first()
+    first_session = Session.query.filter(Session.scheduled).filter_by(project=project).order_by("created_at asc").first()
+    last_session = Session.query.filter(Session.scheduled).filter_by(project=project).order_by("created_at desc").first()
```
Comment (jace) [inline funnel/views/schedule.py]: order_by values should be in single quotes.

File: funnel/views/session.py [hunk 14]
```diff
@@ -17,19 +17,20 @@ def rooms_list(project):
 
 
 def session_form(project, proposal=None, session=None):
+    # Look for any existing unscheduled session
+    if proposal and not session:
+        session = Session.for_proposal(proposal, create=True)
+        db.session.commit()
```
Comment (jace) [inline funnel/views/session.py]: If this is a GET method, we can't create and commit here.

File: migrations/versions/d62b392b0f25_session_start_end_nullable.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+
+def upgrade():
+    op.alter_column('session', 'end', existing_type=postgresql.TIMESTAMP(), nullable=True)
+    op.alter_column('session', 'start', existing_type=postgresql.TIMESTAMP(), nullable=True)
+    op.create_check_constraint(u'ck_session_start_end_nullable', 'session',
```
Comment (jace) [inline migrations/versions/d62b392b0f25_session_start_end_nullable.py]: Name as above.

**Pull request merged**